torchsynth/module.py

```diff
@@ -303,24 +303,29 @@ def _ramp(self, start, duration: T, inverse: bool = False) -> Signal:
         # Shape ramps.
         ramp = ramp - start_[:, None]
         ramp = torch.maximum(ramp, T(0.0))
-        ramp = ramp / duration_[:, None]
+        ramp = (ramp + EPSILON) / (duration_[:, None] + EPSILON)
         ramp = torch.minimum(ramp, T(1.0))
 
         if inverse:
-            ramp = 1 - ramp
+            ramp = 1.0 - ramp
+            ramp[torch.sum(ramp, axis=1) == 0] = 1.0
 
         # Apply scaling factor.
         ramp = torch.pow(ramp, self.p("alpha")[:, None])
 
         return ramp.as_subclass(Signal)
 
-    def make_attack(self) -> Signal:
-        attack = self.p("attack")
+    def make_attack(self, note_on_duration) -> Signal:
+        attack = torch.minimum(self.p("attack"), note_on_duration)
         return self._ramp(torch.zeros(self.batch_size, device=attack.device), attack)
 
-    def make_decay(self) -> Signal:
+    def make_decay(self, note_on_duration) -> Signal:
+        attack = torch.minimum(self.p("attack"), note_on_duration)
+        decay = torch.maximum(note_on_duration - self.p("attack"), T([0.0]))
+        decay = torch.minimum(decay, self.p("decay"))
```

Comment:
Calculations to accommodate a note_hold_on value that might be cutting into the attack/decay phase of the envelope. Not sure how this will differentiate.